spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/messaging/TraceChannelInterceptor.java

```diff
@@ -53,6 +54,9 @@ public TraceChannelInterceptor(BeanFactory beanFactory) {
 	@Override
 	public void afterSendCompletion(Message<?> message, MessageChannel channel,
 			boolean sent, Exception ex) {
+		if (channel instanceof DirectChannel) {
```

Comment:
I think that's not good. The channel might be an AOP proxy, so we don't have an access to the target class.
And also there are many other `AbstractMessageChannel` which maybe tracing candidates as well.

In other words we really should check here for those channel types who are *interceptable*.

In my best feeling that should be like:
```
if (!(channel instanceof InterceptableChannel) && 
          !(channel instanceof ExecutorChannelInterceptorAware) && 
           channel instanceof ChannelInterceptorAware)
```
No, classes to check, only interfaces!
And looks like all the cases are covered.

The `InterceptableChannel` and `ExecutorChannelInterceptorAware` works as is without this fix though.

WDYT?

Reply:
As discussed offline, I've added the check on the proxy of the `DirectChannel` and we came to the conclusion that the only class that makes sense to verify is indeed `DirectChannel` cause the processing will take care in the same JVM process.